src/pygaps/characterisation/whittaker.py

```diff
@@ -0,0 +1,155 @@
+
+    Where :math:`\Delta \lambda` is the adsorption potential, and
+    :math:`\lambda_{p} is the latent heat of the liquid-vapour change at
+    equilibrium pressure.
+
+    Whittaker determined :math:`\Delta \lambda` as;
+
+    .. math::
+        \Delta \lambda = RT\ln{\left[\left(\frac{p^{sat}}{b^{\frac{1}{t}}\right)\left(\frac{\Theta^{t}}{1-\Theta^{t}}\right) \right]}
+
+    Where :math:`p^{sat}` is the saturation pressure, :math:`\Theta` is the
+    fractional coverage, and :math:`b` is derived from the equilbrium constant,
+    :math:`K` as :math:`b = \frac{1}{K^t}`. In the case that the adsorptive is
+    above is supercrictial, the pseudo saturation pressure is used;
+    :math:`p^{sat} = p_c \left(\frac{T}{T_c}\right)^2`. 
+
+    The exponent :math:`t` is only relevant to the Toth version of the method,
+    as for the Langmuir model it reduces to 1. Thus, :math:`\Delta \lambda`
+    becomes
+
+    .. math::
+        \Delta \lambda = RT\ln{\left(\frac{p^{sat}}{b^{\frac{1}{t}}\right)
+    """
+
+    if isotherm.model.name not in ['Langmuir', 'Toth']:
+        raise ParameterError('''Whittaker method requires either a Langmuir or Toth
+                         model isotherm''')
+    else:
+        R = scipy.constants.R
+        n_m = isotherm.model.params['n_m']
+        K = isotherm.model.params['K']
+        T = isotherm.temperature
+        try:
+            p_sat = isotherm.adsorbate.saturation_pressure(temp=298, unit='K')
+        except CalculationError:
+            print(f"{isotherm.adsorbate} does not have saturation pressure "
+                  f"at {isotherm.temperature} K. Calculating psduedo-saturation "
+                  f"pressure...")
+            p_c = isotherm.adsorbate.get_prop('p_critical')
+            T_c = isotherm.adsorbate.get_prop('t_critical')
+            p_sat = p_c * ((T / T_c)**2)
+            p_sat = p_sat * 100  # unit conversion
+
+        if isotherm.model.name == 'Langmuir':
+            t = 1
+        else:
+            t = isotherm.model.params['t'] # equivalent to m in Whittaker
+
+        b = 1 / (K**t)
+
+        df = pd.DataFrame(columns=['Loading', 'q_st'])
+
+        first_bracket = p_sat / (b**(1/t))  # don't need to calculate every time
+        for n in loading:
+            p = isotherm.pressure_at(n) * 1000 
+            sorptive = str(isotherm.adsorbate)
+
+            # check that it is possible to calculate lambda_p
+            if p < PropsSI('PTRIPLE', sorptive) or p > PropsSI('PCRIT', sorptive) or np.isnan(p):
+                pass
+
+            else:
+                lambda_p = heat_vap(p, sorptive)
+
+                theta = n / n_m  # second bracket of d_lambda
+                theta_t = theta**t
+                second_bracket = (theta_t / (1 - theta_t))**((t-1)/t)
+                d_lambda = R * T * np.log(first_bracket * second_bracket)
+
+                q_st = d_lambda + lambda_p + (R*T)
+
+                df = df.append(pd.DataFrame({'Loading': [n],
```

Comment:
DataFrame appending is inherently inefficient as it relies on a full memory copy and reallocation. Better options are: (i) use lists (ii) create the df at required size then allocate during the loop or (iii) use vector-based operations to make loop unnecessary.

Reply:
All resolved in latest commit, except `heat_vap()` function.